Fill gaps limited 7665

Ockenfuss opened the pull request:
# Improve gap filling
- Closes #7665 
- Combine current gap limiting in xarray (max_gap) with capabilities of pandas (limit_direction, limit_area) regarding filling of missing values
- Provide a common function `fill_gaps()` to control gap length for all filling functions (ffill/bfill/interpolate_na/fillna)

This PR involves a full implementation, documentation and corresponding tests. As discussed in #7665, a new API function `fill_gaps` was introduced, to avoid breaking changes of existing code and keep the argument inflation of the filling functions to a minimum.

- [x] Closes #7665 (also #9392 )
- [x] Tests added

TBD if accepted:
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`
- [x] Add Example in Documentation

# Deprecations
- Using `limit` in interpolate_na now requires numbagg or bottleneck. So far, limit worked without bottleneck, max_gap required it (not documented). Reason: max_gap and limit now share the same codebase using ffill. Since no one ever complained about the missing documentation and the error message is pretty clear and easy to resolve ("No module named 'bottleneck"), I hope this might be acceptable.

Ockenfuss committed:
Introduce new arguments limit_direction, limit_area, limit_use coordinate

Ockenfuss committed:
Use internal broadcasting and transpose instead of ones_like

Ockenfuss committed:
Typo: Default False in doc for limit_use_coordinates

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Ockenfuss committed:
Towards masked implementation

Ockenfuss committed:
Working fill_gaps implementation

Ockenfuss committed:
Remove keep_attrs from docstring of filling functions

Ockenfuss committed:
Fix typos, undo empty spaces, remove temporarily introduced arguments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

max-sixty commented on xarray/tests/test_missing.py:
```diff
@@ -559,6 +633,114 @@ def test_bfill_dataset(ds):
+
+
+def test_get_gap_right_edge():
+    n = np.nan
+    arr = [
+        [n, 1, n, n, n, n, n, n, 4],
+        [n, n, n, 1, n, n, 4, n, n],
+    ]
```
(FWIW this is a really nice way of aligning arrays, haven't thought of this before)

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -6864,6 +6877,174 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
+            | datetime.timedelta
+        ) = None,
+        limit_direction: LimitDirectionOptions = "both",
+        limit_area: LimitAreaOptions | None = None,
```
Refine-y but:
- are `interpolate` and `extrapolate` clearer options?
- is there a better name for the param? I'm not sure. "area" sounds 2D to me

Ockenfuss replied:
True. I adopted this nomenclature from pandas. What about `limit_region`, does this sound less 2D?
Personally, I like the 'inside'/'outside' options, 'interpolate' sounds close to linear interpolation to me...

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -6864,6 +6877,174 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
+        ) = None,
+    ) -> GapMask:
+        """Fill in gaps in the data using one of several filling methods.
+        Allows for fine control on how far to extend the valid data into the gaps and the maximum size of the gaps to fill.
```
Could we clearly list the methods here on the returned object? (Or lmk if I'm missing them — they're in the doctest but otherwise I don't see them, thanks)

max-sixty commented on xarray/core/dataset.py:
```diff
@@ -6723,6 +6731,9 @@ def interpolate_na(
         from xarray.core.missing import _apply_over_vars_with_dim, interp_na
 
+        if keep_attrs is None:
+            keep_attrs = _get_keep_attrs(default=False)
```
```suggestion
            keep_attrs = _get_keep_attrs(default=True)
```

IIRC we're gradually moving in this direction

Ockenfuss replied:
Done! Also, I noticed ffill/bfill have no keep_attrs, maybe this is a separate PR

max-sixty replied:
(Yes and actually I think we're trying to remove `keep_attrs` gradually, so fine to include it if you've already done that, but let's not add it anywhere.)

max-sixty commented on xarray/core/dataarray.py:
```diff
@@ -3766,6 +3769,160 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
 
         return bfill(self, dim, limit=limit)
 
+    def fill_gaps(
```
Any thoughts from others on the naming? Would `.fill` be insufficiently specific that it's filling `na`? Would `fill_missing` be clearer than `fill_gaps`?

Ockenfuss replied:
Open to any of those. `.fill` sounds very concise, but maybe this is easily confused with `.ffill`

max-sixty replied:
I think `.fill` could be quite nice, do others have a view?

dcherian replied:
Maybe `gap_filler` instead, since this method does not actually fill the gaps.

I'm also wondering if its better to have a method that constructs the appropriate mask that can be used later
```
mask = ds.gap_mask(max_gap=...)
ds.ffill(...).where(~mask)
```

Ockenfuss replied:
`gap_filler` emphasizes the returned object type nicely! However, I choose `fill_gaps` because it fits the naming scheme of other object-returning functions better (e.g. `rolling` and `coarsen` are not called `roller` and `coarser` in xarray, even though the operation is not perfomed immediately and an object is returned).
Since I am a non-native speaker, I will follow your recommendations regarding nomenclature.

The function API is also presented as an alternative in the [initial proposal](https://github.com/pydata/xarray/issues/7665#issuecomment-1995178933). I decided to go for the object way because it is shorter (one line) and less error prone (you might easily forget the `~`). If the mask is required, you can easily get it from the object:
```python
mask=ds.gap_filler(...).mask
```

max-sixty commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
-            )
-        nan_block_lengths = _get_nan_block_lengths(self, dim, index)
-        arr = arr.where(nan_block_lengths <= max_gap)
+class GapMask:
```
Refine-y but re the name: it's more of a `GapFiller` than a `GapMask`, no? I've thought of `mask` as something that removes valid values.

(or just `Filler` or `Fill`?)

Ockenfuss replied:
Good point! I would change this accordingly, once we decided for a name for the `.fil_gaps` method.

max-sixty commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+        method.__doc__ = _FILL_MISSING_DOCSTRING_TEMPLATE.format(name=name)
+        return method
+
+    ffill = _fill_method("ffill")
+    bfill = _fill_method("bfill")
+    fillna = _fill_method("fillna")
+    interpolate_na = _fill_method("interpolate_na")
```
Maybe not worth new names, but the naming would be more orthogonal to just have `.forward`, `.backward` etc

Ockenfuss replied:
In this case, all names could be adapted: Instead of `fill_gaps().ffill/bfill/fillna/interpolate_na`, we could call them `fill_gaps().forward/backward/constant/interpolate` to avoid the repeated "fill" terms. However, I am not sure if this is worth it to deviate from the existing xarray nomenclature.

max-sixty commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+        method.__doc__ = _FILL_MISSING_DOCSTRING_TEMPLATE.format(name=name)
+        return method
+
+    ffill = _fill_method("ffill")
```
Does this actually enforce `direction="forward"`?

Ockenfuss replied:
No, see the example above.

max-sixty reviewed (approved):
This seems extremely good!!

I left some refine-y comments, nothing preclusive.

What do others think? I don't use `interpolate_na` myself much so other may have views...

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+        method.__doc__ = _FILL_MISSING_DOCSTRING_TEMPLATE.format(name=name)
+        return method
+
+    ffill = _fill_method("ffill")
```
This is a nice way of dynamically adding these methods but from a static typing point of view that's always annoying because the arguments cannot be resolved. 

Since it is only a handful of methods maybe it is better to add them manually?

Ockenfuss replied:
I "stole" this pattern from the rolling implementation :) But true, manual adding also allows for more flexibility. I changed this accordingly.

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
-            )
-        nan_block_lengths = _get_nan_block_lengths(self, dim, index)
-        arr = arr.where(nan_block_lengths <= max_gap)
+class GapMask:
```
This should be a generic class: 

```suggestion
class GapMask(Generic[T_Xarray]):
    content: T_Xarray
    mask: np.ndarray
```

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+        self.mask = mask
 
+    def _fill_method(name: str, _fill_function: Callable | None = None):
+        def method(self, *args, _fill_function=_fill_function, **kwargs):
```
```suggestion
        def method(self, *args, _fill_function=_fill_function, **kwargs) -> T_Xarray:
```

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+        self.content = content
+        self.mask = mask
 
+    def _fill_method(name: str, _fill_function: Callable | None = None):
```
```suggestion
    def _fill_method(name: str, _fill_function: Callable | None = None) -> Callable[..., T_Xarray]:
```
Or whatever the arguments are

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+
+
+def mask_gaps(
+    self,
```
```suggestion
    self: T_Xarray,
```
Since this is not a member method of any class, consider naming this `obj` instead.

Ockenfuss replied:
True, I renamed all self outside classes!

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -365,26 +515,112 @@ def interp_na(
+    limit_direction: LimitDirectionOptions = "both",
+    limit_area: LimitAreaOptions | None = None,
+    max_gap: int | float | str | pd.Timedelta | np.timedelta64 | dt.timedelta = None,
+) -> GapMask:
```
```suggestion
) -> GapMask[T_Xarray]:
```

headtr1ck commented on xarray/core/dataarray.py:
```diff
@@ -3476,10 +3479,10 @@ def fillna(self, value: Any) -> Self:
     def interpolate_na(
         self,
-        dim: Hashable | None = None,
+        dim: Hashable,
```
This seems like a breaking change, but I'm not sure what this does when passing None...

Ockenfuss replied:
Currently, it raises a `NotImplementedError`. Actually, this was never working, someone also raised an issue recently: #9392

headtr1ck reviewed (commented):
Basically all the new methods are missing return types, please consider adding them. 

The new class should be generic, I have added a few review comments here and there but more adoptions are needed.

Ockenfuss committed:
Add line break for readability

Ockenfuss commented on xarray/core/dataarray.py:
```diff
@@ -3766,6 +3769,160 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
+        array([nan,  2., nan, nan,  5.,  5.,  0.])
+        Coordinates:
+        * x        (x) int64 56B 0 1 2 3 4 5 6
+        >>> da.fill_gaps(dim="x", limit_area="inside").fillna(9)
```
Of course, done!

Ockenfuss commented on xarray/core/dataset.py:
```diff
@@ -6864,6 +6877,174 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
+            For ``limit_direction=forward``, distances are ``[nan, nan, nan, 0, 1, 2, 0, 1, 2]``.
+            To only fill gaps less than a given length,
+            see ``max_gap``.
+        limit_direction: {"forward", "backward", "both"}, default: "forward"
+            Consecutive NaNs will be filled in this direction.
```
I think it is actually useful to have this parameter and allow to set it independently of the filing function. Then, for example, you can do things like:
```python
n = np.nan
test=xr.DataArray([1,1, n, n, n, 2, 2], dims=["x"])
test.fill_gaps('x', limit=1, limit_direction='backward').ffill('x')
```
which effectively pulls the gap left edge towards the gap right edge:
```
<xarray.DataArray (x: 7)> Size: 56B
array([ 1.,  1., nan, nan,  1.,  2.,  2.])
Dimensions without coordinates: x
```
I think there might be use cases like for this, constructing masks or advanced indexing arrays maybe.

max-sixty replied:
Hmmm, in my (probably naive) mental model of filling, this wouldn't work like this. How would we explain in words why we get the additional `1`?

Ockenfuss replied:
I see your point... My view on this:
- `fill_gaps()` decides **which** values to fill (here 1 step backward, i.e. the right edge of every gap)
- The following function decides **how** they are filled (`ffill` here)

Both are two independent steps.
Maybe it is clearer to use 'left'/'right' instead of 'forward'/'backward' terminology?
```
limit_side: {"left", "right", "both"}, default: "both"
    Whether to limit the filling to the left, right, or both sides of the gap
     - "left": Fill only the left (lower index) side of each gap.
     - "right": Fill only the right (higher index) side of each gap.
     - "both": Fill from both sides of each gap.
```

Ockenfuss replied:
I could also join this Wednesday developer meeting to get feedback on naming/API things.

max-sixty replied:
I see what you mean. I personally think that's quite confusing, but 50% chance it's just me.

Do others have a view?

I'm thinking that filling is simpler than this: it starts at the edge of a gap — and then fills the gap from the edge, which is by definition backwards or forwards (because it's already at an edge so can only go one direction)...

max-sixty replied:
(yes re Wednesday calls! Unfortunately it clashes with something I have so I find it tough to go, but others will be there!)

Ockenfuss replied:
From a technical point of view, it is also not easy to remove the limit_direction argument in case of ffill/bfill, since fill_gaps is a two stage process. First, the mask is created via `fill_gaps(limit_direction='forward')` and a GapMask object is returned. Here, we don't know yet which filling function the user will apply later.

Therefore, and since the code itself is safe and well-defined, I decided to leave it as is for now. The default for limit_direction is 'both' anyway, so the counterintuitive forward/bfill (backward/ffill) cases must be triggered intentionally.

Ockenfuss commented on xarray/core/dataset.py:
```diff
@@ -6864,6 +6877,174 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
+
+            - False: a consecutive integer index is created along ``dim`` (0, 1, 2, ...).
+            - True: the IndexVariable `dim` is used.
+            - String: specifies the name of a coordinate variable to use as the index.
```
You mean the last "String" option, I guess. True, the only use case of passing dim and the coordinate separately could be if the coordinate is multidimensional. However, currently this is not supported:

https://github.com/pydata/xarray/blob/1626489d4ae08449880f6c3605ba43f0f05d9805/xarray/core/missing.py#L412C1-L416C10

max-sixty replied:
That's a good point though — we could (in the future) take the values from a 2D coord is a single direction. So leaving this as-is seems reasonable to me...

Ockenfuss committed:
Enforce kwargs to be passed by name

Ockenfuss commented on xarray/core/dataset.py:
```diff
@@ -6864,6 +6877,174 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
 
+    def fill_gaps(
+        self,
+        dim: Hashable,
+        use_coordinate: bool | Hashable = True,
```
Adapted in DataArray/Dataset!

Ockenfuss committed:
Keep_Attrs: Default to True

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Ockenfuss commented:
> Basically all the new methods are missing return types, please consider adding them.
> 
> The new class should be generic, I have added a few review comments here and there but more adoptions are needed.

Thanks a lot for the review of the typing! I`ll work through them asap and try to make all adoptions (probabily still need some help afterwards, though ...)

Ockenfuss committed:
Explicitly add fill functions in GapMask object
- Allows for more flexibility, e.g. optional dim arguments
- Better static typing

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Ockenfuss committed:
Add type hints to most arguments, return types

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Ockenfuss committed:
Fix accidental double pasting of arguments

Ockenfuss committed:
Fix more mypy errors

Ockenfuss committed:
Bottleneck is required for limit functionality

max-sixty commented on xarray/core/dataarray.py:
```diff
@@ -3776,6 +3776,8 @@ def fill_gaps(
         """Fill in gaps in the data using one of several filling methods.
         Allows for fine control on how far to extend the valid data into the gaps and the maximum size of the gaps to fill.
 
+        *Requires bottleneck.*
```
I think numbagg should work too? (And is preferred so should be listed first)

Ockenfuss replied:
True! Also changed for ffill/bfill docs

Ockenfuss committed:
Docs: Require numbagg or bottleneck for ffill/bfill/fill_gaps

Ockenfuss committed:
Rework index conversion to have consistent typing

headtr1ck commented:
> @headtr1ck I did my best to add type hints everywhere. Not sure how to resolve the last four remaining mypy errors...

I'm currently on holiday and it's difficult to debug these errors on the phone.
I can do it in 1-2 weeks or someone else can help out?

headtr1ck commented on xarray/core/missing.py:
```diff
@@ -20,44 +18,207 @@
+    # Calculate individual masks
+    masks = []
+    if need_limit_mask:
+        masks.append(
```
I think it is difficult to tell mypy that limit cannot be a Hashable here do to the dynamic way of how it is handled here.
Maybe the best course of action is to add an assertion (not sure which exactly, maybe `assert not utils.is_hashable(limit)`, but maybe this is too restrictive?)
Easier might be to add a `#type: ignore[arg-type]` and a little comment as to why.

max-sixty commented:
There are a couple of outstanding questions about the interfaces; in particular:
- forward vs back vs direction: https://github.com/pydata/xarray/pull/9402#discussion_r1730481171
- some names: https://github.com/pydata/xarray/pull/9402#discussion_r1742677060

Re typing — are the interfaces typed correctly? I _think_ they are. Assuming those are correct, I would vote that it's fine to add ignores internally (ofc the internal checks are useful tests for the external interfaces, so ideally we'd fix those)

What's the best way of resolving those outstanding questions @pydata/xarray ? I know I've been terrible on attending calls now I'm on PT, but maybe we could do it on one of those? (would be great to minimize the delay on these sorts of PRs — I think they're really valuable and would love to engender more...)

Ockenfuss commented:
Nothing else, I think, that is it! I am on conference this week, but mostly available the next two weeks, if anyone wants to discuss. Otherwise, I will join the next dev meeting on Sep. 25.

max-sixty commented:
Did you manage to discuss this this morning?

kmuehlbauer commented:
> Did you manage to discuss this this morning?

@max-sixty Yes, @Ockenfuss explained what decisions are to be made and encouraged to have a look here.

Ockenfuss commented:
Hi @max-sixty, not too much community feedback yet :) What do you think, shall we have another look on the three open naming issues and then push this over the finish line?
I still think this is quite useful functionality for all experimentalists, where gaps of all sizes are common in measurement data.

max-sixty commented:
> What do you think, shall we have another look on the three open naming issues and then push this over the finish line?

Yes let's do it! I would bias for action; if folks disagree then dissent welcome.

What's remaining? Personally I think we should simplify this: https://github.com/pydata/xarray/pull/9402#discussion_r1730481171 + resolve any names. The _external_ types should be correct; internally some ignores are OK...

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Ockenfuss committed:
Remove typo (double line)

Ockenfuss committed:
Add a hint in the interpolate_na docs about the forward (legacy) behaviour.

Ockenfuss committed:
Add example in User guide

Ockenfuss committed:
Fix typo in documentation

Ockenfuss committed:
Fix typing errors (ignore types for limit argument internally)

Ockenfuss committed:
Remove return type of interp_na to avoid mypy error

Ockenfuss committed:
Include documentation for GapMask Object

Ockenfuss committed:
Include references in docs between filling functions

Ockenfuss committed:
Doc-Bug: Default for direction is both

Ockenfuss commented:
Ok, I think from my side this is ready for merge. I stayed with the fill_gaps and limit_area names for now. Check out the doc build pages for fill_gaps and the GapMask object to get a feeling for it and let me know if it feels right :)

This PR leaves all existing fill function API untouched. If we want to bring some of the new limit arguments to the existing functions (like DataArray.fillna(limit=...)), this could be done later in separate PRs.

TBD:
- Update Whatsnew

max-sixty commented on xarray/core/dataarray.py:
```diff
@@ -3810,6 +3823,153 @@ def bfill(self, dim: Hashable, limit: int | None = None) -> Self:
+
+        Notes
+        -----
+        ``Limit`` and ``max_gap`` have different effects on gaps: If ``limit`` is set, *some* values in a gap will be filled (up to the given distance from the boundaries). ``max_gap`` will prevent *any* filling for gaps larger than the given distance.
```
```suggestion
        ``limit`` and ``max_gap`` have different effects on gaps: If ``limit`` is set, *some* values in a gap will be filled (up to the given distance from the boundaries). ``max_gap`` will prevent *any* filling for gaps larger than the given distance.
```

max-sixty commented:
Looks great, I think this is indeed very close to landing.

The one thing I'm stuck on is the concept of `fill_direction="backward" ... .ffill...` — can we explain what the purpose of this is? My mental model is that we're filling missing data from an edge in a single direction; I don't understand how what's going on to fill backwards in a forward fill.

(Totally fine if others get this, I'm not the gatekeeper here, but do think it's important the API makes sense...)

Ockenfuss committed:
Typo in Documentation

Co-authored-by: Maximilian Roos <5635139+max-sixty@users.noreply.github.com>

Ockenfuss commented:
See [this comment](https://github.com/pydata/xarray/pull/9402#discussion_r1730428987) for an explanation, or just try  out what happens :) For example:
```python
da=xr.DataArray([1, np.nan, np.nan,  2], dims=['x'])
da.fill_gaps('x', limit=1, limit_direction='backward').ffill()
```
```
<xarray.DataArray (x: 4)> Size: 32B
array([ 1., nan,  1.,  2.])
Dimensions without coordinates: x
```
I agree, this is not maybe obvious. [Here](https://github.com/pydata/xarray/pull/9402#discussion_r1925677551), I explained why I left it as is. If you want, I can try to keep track of the limit_direction argument upon mask creation and raise an exception, if the user applies ffill later. But this requires some tweaks in the code.

Ockenfuss committed:
Do not allow further limit or max_gap specification when calling interpolate_na on a GapMask object

Ockenfuss committed:
Make two stages of filling clear in fill_gaps documentation

max-sixty commented:
Yes, I see what it does, but (very very respectfully, with much admiration for the overall PR!), it doesn't make semantic sense to me.

When I think "backward", I mean "look behind you, is there a gap? OK now fill it". I don't see how it can mean "start _one_ step back, and then decide which direction to continue". Why only one step?

I think we should try hard to have xarray's interfaces be understandable without much technical knowledge. (Indeed, I think there's probably a more precise way of describing my objection around edges and continuous vs. discrete, but hopefully this suffices). 

Would you object to removing the `limit_direction` param, and deferring that to the method?

Ockenfuss commented:
Thank you very much for the feedback and efforts! I am not entirely sure what you mean with decide and continuous vs discrete, I have to admit. I drew you a little picture, to visualize how I see those things :)
![1000039597](https://github.com/user-attachments/assets/3c979018-5f16-4594-9304-93d9329eb71d)


But I think, I can rework the API. Just to clarify before I start, what you aim for is the following?
```python
da.fill_gaps('x', limit=1).ffill() # no direction argument anymore in the first function
```
In the case of interpolate_na and fillna, which work in all directions:
```python
da.fill_gaps('x', limit=1).interpolate_na(limit_direction='forward')
```

max-sixty commented:
Ha, thanks a lot for the drawing! I appreciate it!

FWIW the step there that didn't make sense to me is "one step backward" for `bfill` & `ffill` — my mental model is that filling doesn't require this, and if "one step backward" isn't in others' mental models, they'll also be confused by it...

`interpolate_na` is an interesting one, since you're correct that can go both ways: 
- Having a kwarg on `interpolate_na` is reasonable. 
- I could imagine a builder pattern where we just keep running methods `.interpolate(...).with_direction(...)`, but that's not really pythonic and is inconsistent with xarray's API
- or have `limit_direction` as it is now, but default to the fill direction and raise an error if it's not the same

Ockenfuss committed:
Default to forward for ffill and backward to bfill.
Raise error if this is not the case.
This implies that the mask creation needs to be delayed until the filling stage.

Ockenfuss committed:
Update api.rst and GapMask Attributes

Ockenfuss commented:
I went for option 3: Default to 'forward' in case of ffill and 'backward' in case of bfill. If this is not the case, an error is raised. (This is similar to what pandas is doing, check out the docstring [here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html).)
I adapted the code and docs and added some tests to check this behaviour. What do you think?

Ockenfuss committed:
Split ffill/bfill pandas compatibility test into separate test method

Ockenfuss committed:
Fix doc examples and correct return type hint

Ockenfuss committed:
Require bottleneck for ffill test

Ockenfuss committed:
Fix mask type hint in two positions.

Ockenfuss committed:
Remove fill_value in pandas when method=ffill/bfill

Ockenfuss committed:
Remove ffill check against old pandas version

max-sixty commented on xarray/tests/test_missing.py:
```diff
@@ -1131,6 +1143,25 @@ def test_fill_gaps_limit():
     coords = {"yt": ("y", times)}
     da = xr.DataArray([n, n, 2, n, n, 5, n, n], dims=["y"], coords=coords)
 
+    mask = da.fill_gaps(dim="y", limit_direction="backward")
```
Can we add a test for not passing a direction? That seems like the most likely case

Ockenfuss replied:
Good point, done! See also the tests in test_mask_gap_limit_2d.

Ockenfuss committed:
Add additional tests for the direction kwarg in combination with ffill/bfill

keewis commented:
I'm late to the discussion (thanks for sticking with us for so long, even though we've been really, really slow!), but to me `ffill` with `limit_direction="backwards"` seems to mean "take the value _before_ the gap, but start filling (and counting) from the back" (and the inverse for `bfill`: "take the values from after the gap and start filling from the front of the gap"). Which I also find odd, because I can't imagine a use case for that (maybe there is one I just can't see?) and we can't use `pandas` as a precedent; that only has `limit_area` which is a different thing. I do get the value of `limit_direction` for `interpolate`, though.

Other than that, the only suggestion I have is to look at the naming of the API (I guess I'm bikeshedding?):
```python
ds.fill_gaps(...).interpolate(...)  # no need to call it `interpolate_na` since it already is in a namespace?
```
kinda works, but
```python
ds.fill_gaps(...).ffill(...)
ds.fill_gaps(...).bfill(...)
```
repeats "fill" and feels a bit awkward (it does fit with the `limit*` kwargs, though). What about just `gaps` or, if we don't want to assume everyone understands `gaps` the same way, `na`? That would make the API:
```python
ds.gaps(...).ffill(...)
ds.gaps(...).interpolate(...)
```
or
```python
ds.na(...).ffill(...)
ds.na(...).interpolate(...)
```
The latter may be a bit short and missing the semantic hint of "we're filling gaps", though. (And either way, these are suggestions, not requests, so feel free to reject them if it doesn't make too much sense or if it has already been discussed in a previous version of this PR).

Ockenfuss commented:
Thanks you, no worries, all feedback is welcome everytime!!
### ffill+direction backward issue
Your understanding is correct. I have no immediate use case for it, but I thought initially why forbidding something that is possible. However, Max also thought this case may be confusing, so we decided to raise an error if attempting it now.
### Names
I am open to anything! I am with you that having "gap" somehow in the name helps for the understanding. Just "gaps" is also fine for me.
### Anything else
You might wonder: Why a new fill_gaps API function and not just equip the existing functions with new arguments (e.g. for ffill: limit, limit_area, max_gap, limit_use_coordinate)?
Originally, we were mostly concerened about interpolate_na. Stephan and Deepak recommended a new API, since the existing interpolate_na has a very specific legacy behaviour. So for interpolate_na, we need a new API if we want to avoid deprecations. For ffill/bfill, both solutions have something in favour: The current implementation keeps all gap options nicely in one place. Directly adding the arguments instead would allow for some more freedom, e.g. the limit_direction argument could simply be omitted.

max-sixty commented:
Great!

For the name, a verb is consistent with the existing interface (`resample`, `groupby`, etc). `fill_gaps` seems quite reasonable, and I can't think of anything that's a single word...

...unless `fill` could work? It contends with `np.fill` being more general. But then the methods on the object are specific to filling missing values.

...so maybe that could work? If others agree then I would be in favor of `fill`...

---

Unless we do decide to change the name, is anything else remaining? Last call for feedback @pydata/xarray !

kmuehlbauer commented:
How about `gapfill`?  Two characters less than `fill_gaps`, single word and verb.

Ockenfuss commented:
gapfill sounds good to me! @max-sixty, are you ok as well? Then I change code+docs accordingly.

max-sixty commented:
> How about `gapfill`? Two characters less than `fill_gaps`, single word and verb.

Does `gapfill` go against the python convention of separating words with underscores?

(I don't want to bikeshed, if others are really keen on this then totally OK)

@pydata/xarray any final call for name suggestions?

Ockenfuss commented:
To summarize:
- fill: Pro: short+straightforward. Con: Easily confused with ffill, does not highlight the 'gap'
- fill_gaps: Pro: Highlights the gaps. Con: Long
- gapfill: Pro: Shorter and easier to type than gapfill. Con: nount first and no underscore separation are not pythonic.

I have a weak tendency towards fill_gaps (that's why I chose it in the beginning). But your view as external reviewer is probably more objective. I think you have to be the benevolent dictator and make a decision :)

Ockenfuss committed:
Add a small comment regarding the original behaviour of inter_na

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Ockenfuss committed:
Using the limit argument requires bottleneck (since ffill/bfill are used)

Ockenfuss committed:
Add GapMask object to API docs
- adapt jupyter/ipython command to new notation in user guide

keewis committed:
Merge branch 'main' into fill_gaps_limited_7665

kmuehlbauer committed:
Merge branch 'main' into fill_gaps_limited_7665
